### ISSUE TITLE
[BUG/ISSUE] Deferred-length character component 'key' at (1) is not yet supported

### ISSUE THREAD

**Issue by FeiYao-Edinburgh** [geoschem/geos-chem#464]
Hi there,

I recently started to use GC 12.9.3, but soon encountered the compilation error. My steps are:

- Download the source code and UT:
```
 # Download the GEOS-Chem source code
 git clone https://github.com/geoschem/geos-chem Code.12.9.3
 cd Code.12.9.3
 
 # Create and checkout a new branch "GC_12.9.3" at the tag "12.9.3"
 git checkout -b GC_12.9.3 12.9.3
 cd .. 

 # Download the GEOS-Chem Unit Tester, from which run directories can be generated
 git clone https://github.com/geoschem/geos-chem-unittest UT
 cd UT
 
 # Create and checkout a new branch "GC_12.9.3" at the tag "12.9.3"
 git checkout -b GC_12.9.3 12.9.3
```
- Modify /UT/perl/CopyRunDirs.input properly to create the default geosfp_4x5_standard rundir
- In geosfp_4x5_standard rundir, execute `make -j4 build` (GNU Fortran (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)), which gives the following compilation error
```
make[5]: Entering directory `/exports/csce/datastore/geos/users/s1855106/PhD-Edinburgh/Code.12.8.0/Headers'
gfortran -cpp -w -std=legacy -fautomatic -fno-align-commons -fconvert=big-endian -fno-range-check -O3 -funroll-loops -fopenmp -mcmodel=medium -fbacktrace -g -DMODEL_CLASSIC -DLINUX_GFORTRAN -DNC_HAS_COMPRESSION -DUSE_REAL8 -J../mod -I/usr/include -I/usr/include -c -ffree-form -ffree-line-length-none charpak_mod.F90
gfortran -cpp -w -std=legacy -fautomatic -fno-align-commons -fconvert=big-endian -fno-range-check -O3 -funroll-loops -fopenmp -mcmodel=medium -fbacktrace -g -DMODEL_CLASSIC -DLINUX_GFORTRAN -DNC_HAS_COMPRESSION -DUSE_REAL8 -J../mod -I/usr/include -I/usr/include -c -ffree-form -ffree-line-length-none inquireMod.F90
gfortran -cpp -w -std=legacy -fautomatic -fno-align-commons -fconvert=big-endian -fno-range-check -O3 -funroll-loops -fopenmp -mcmodel=medium -fbacktrace -g -DMODEL_CLASSIC -DLINUX_GFORTRAN -DNC_HAS_COMPRESSION -DUSE_REAL8 -J../mod -I/usr/include -I/usr/include -c -ffree-form -ffree-line-length-none dictionary_m.F90
gfortran -cpp -w -std=legacy -fautomatic -fno-align-commons -fconvert=big-endian -fno-range-check -O3 -funroll-loops -fopenmp -mcmodel=medium -fbacktrace -g -DMODEL_CLASSIC -DLINUX_GFORTRAN -DNC_HAS_COMPRESSION -DUSE_REAL8 -J../mod -I/usr/include -I/usr/include -c -ffree-form -ffree-line-length-none precision_mod.F90
dictionary_m.F90:26.41:

     character(len=:), allocatable :: key
                                         1
Error: Deferred-length character component 'key' at (1) is not yet supported
dictionary_m.F90:26.41:

     character(len=:), allocatable :: key
                                         1
Error: Deferred-length character component 'key' at (1) is not yet supported
dictionary_m.F90:26.41:

     character(len=:), allocatable :: key
                                         1
Error: Deferred-length character component 'key' at (1) is not yet supported
dictionary_m.F90:138.16:

       allocate(this%buckets(h)%entries, source=tmp_bucket%entries)
                1
Error: Array specification required in ALLOCATE statement at (1)
```
Clearly, the error occurred in Headers/dictionary_m.F90, but I have no idea to fix it.

I have also explored 12.9.2, 12.9.0, and 12.8.0. All returned the issues above.

BTW I have configured necessary computational environment following [this](http://wiki.seas.harvard.edu/geos-chem/index.php/Configuring_your_computational_environment).

But it seems more like a FORTRAN problem? Any help would be appreciated!

**Comment by FeiYao-Edinburgh** [comment]
I addressed the issue by switching from
```
export FC=gfortran
export CC=gcc
export CXX=g++
```
to
```
export FC=ifort
export CC=icc
export CXX=icpc
```
but do not clearly know why...

**Comment by yantosca** [comment]
Thanks for writing.  Your Gfortran compiler is too old.  GEOS-Chem will no longer support gfortran4.  Better to use Gfortran 7 or higher.

**Comment by FeiYao-Edinburgh** [comment]
Thanks for letting know!

**Comment by yantosca** [comment]
I can close this issue out.  Thanks.

**Comment by guapi-hub** [comment]
Hello, could you please let me know if your issue has been resolved? I've upgraded from gcc 4.8 to gcc 11, but the problem still hasn't been solved. I'm using the version gcclassic-14.4.2. The error message is as follows:

/root/gc_4x5_merra2_fullchem_test/CodeDir/src/GEOS-Chem/Headers/dictionary_m.F90:26.41:

     character(len=:), allocatable :: key
                                         1
Error: Deferred-length character component 'key' at (1) is not yet supported
/root/gc_4x5_merra2_fullchem_test/CodeDir/src/GEOS-Chem/Headers/dictionary_m.F90:26.41:

     character(len=:), allocatable :: key
                                         1
Error: Deferred-length character component 'key' at (1) is not yet supported
/root/gc_4x5_merra2_fullchem_test/CodeDir/src/GEOS-Chem/Headers/dictionary_m.F90:26.41:

     character(len=:), allocatable :: key
                                         1
Error: Deferred-length character component 'key' at (1) is not yet supported
/root/gc_4x5_merra2_fullchem_test/CodeDir/src/GEOS-Chem/Headers/dictionary_m.F90:138.16:

       allocate(this%buckets(h)%entries, source=tmp_bucket%entries)
                1
Error: Array specification required in ALLOCATE statement at (1)
make[2]: *** [src/GEOS-Chem/Headers/CMakeFiles/Headers.dir/build.make:140: src/GEOS-Chem/Headers/CMakeFiles/Headers.dir/dictionary_m.F90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1116: src/GEOS-Chem/Headers/CMakeFiles/Headers.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

**Comment by guapi-hub** [comment]
Hello, could you please let me know if your issue has been resolved? I've upgraded from gcc 4.8 to gcc 11, but the problem still hasn't been solved. I'm using the version gcclassic-14.4.2. The error message is as follows:

Error: Deferred-length character component 'key' at (1) is not yet supported
/root/gc_4x5_merra2_fullchem_test/CodeDir/src/GEOS-Chem/Headers/dictionary_m.F90:26.41:

**Comment by FeiYao-Edinburgh** [comment]
Hi @guapi-hub it is clearly an issue with the compiler version again; however, I can only confirm that my GNU Fortran (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0 works with GCClassic v14.1.1.
